File: backend/app/schemas/partner.py
import uuid
from typing import Optional, List
from datetime import datetime
from pydantic import BaseModel, ConfigDict, EmailStr

# --- Contact Schemas ---
class PartnerContactBase(BaseModel):
    name: str
    position: Optional[str] = None
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    is_primary: bool = False

class PartnerContactCreate(PartnerContactBase):
    pass

class PartnerContact(PartnerContactBase):
    contact_id: uuid.UUID
    partner_id: uuid.UUID
    model_config = ConfigDict(from_attributes=True)

# --- Partner Schemas ---
class PartnerBase(BaseModel):
    name: str
    partner_cnc: Optional[str] = None
    type_id: Optional[str] = None
    status_id: Optional[str] = None
    group_id: Optional[str] = None
    area_id: Optional[str] = None
    sub_area_id: Optional[str] = None
    version_id: Optional[str] = None
    imp_type_id: Optional[str] = None
    
    stars: Optional[int] = None
    rooms: Optional[int] = None
    outlets: Optional[int] = None
    address: Optional[str] = None
    system_live_at: Optional[datetime] = None
    last_visit_at: Optional[datetime] = None
    last_visit_type: Optional[str] = None
    last_project: Optional[datetime] = None
    last_project_type: Optional[str] = None

class PartnerCreate(PartnerBase):
    contacts: Optional[List[PartnerContactCreate]] = []

class PartnerUpdate(BaseModel):
    name: Optional[str] = None
    partner_cnc: Optional[str] = None
    type_id: Optional[str] = None
    status_id: Optional[str] = None
    group_id: Optional[str] = None
    area_id: Optional[str] = None
    sub_area_id: Optional[str] = None
    version_id: Optional[str] = None
    imp_type_id: Optional[str] = None
    stars: Optional[int] = None
    rooms: Optional[int] = None
    outlets: Optional[int] = None
    address: Optional[str] = None
    system_live_at: Optional[datetime] = None
    is_active: Optional[bool] = None
    last_visit_at: Optional[datetime] = None
    last_visit_type: Optional[str] = None
    last_project: Optional[datetime] = None
    last_project_type: Optional[str] = None
    contacts: Optional[List[PartnerContactBase]] = None

class Partner(PartnerBase):
    partner_id: uuid.UUID
    created_at: datetime
    updated_at: datetime

    contacts: List[PartnerContact] = []
    
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/store/partner.ts
import { defineStore } from 'pinia'
import apiClient from '../api/api-client'

export interface Partner {
  partner_id: string
  name: string
  partner_cnc: string
  type_id: string
  status_id: string
  stars: number
  rooms: number
  outlets: number
  address: string
  area_id?: string
  sub_area_id?: string
  last_visit_at?: string
  last_visit_type?: string
  last_project?: string
  last_project_type?: string
  version_id?: string
  imp_type_id?: string
  server_information_id?: string
  server_information_detail?: string
  system_live_at?: string
  contacts: any[]
  updated_at: string
}

interface PartnerState {
  partners: Partner[]
  lookups: {
    types: any[]
    statuses: any[]
    groups: any[]
    areas: any[]
    sub_areas: any[]
    versions: any[]
    imp_types: any[]
    server_info: any[]
  }
  isLoading: boolean
  error: string | null
}

export const usePartnerStore = defineStore('partner', {
  state: (): PartnerState => ({
    partners: [],
    lookups: {
      types: [],
      statuses: [],
      groups: [],
      areas: [],
      sub_areas: [],
      versions: [],
      imp_types: [],
      server_info: [],
    },
    isLoading: false,
    error: null,
  }),

  actions: {
    async fetchPartners() {
      this.isLoading = true
      try {
        const response = await apiClient.get('/partners/')
        this.partners = response.data
      } catch (err: any) {
        this.error = 'Gagal memuat data partner.'
      } finally {
        this.isLoading = false
      }
    },

    async fetchLookups() {
      try {
        const response = await apiClient.get('/lookups/partner')
        this.lookups = response.data
      } catch (err: any) {
        console.error('Failed to fetch lookups', err)
      }
    },

    async createPartner(payload: any) {
      try {
        const response = await apiClient.post('/partners/', payload)
        this.partners.unshift(response.data)
        return response.data
      } catch (err: any) {
        throw err
      }
    },

    async updatePartner(id: string, payload: any) {
      try {
        const response = await apiClient.patch(`/partners/${id}`, payload)
        const index = this.partners.findIndex(p => p.partner_id === id)
        if (index !== -1) this.partners[index] = response.data
        return response.data
      } catch (err: any) {
        throw err
      }
    },

    async deletePartner(id: string) {
      try {
        await apiClient.delete(`/partners/${id}`)
        this.partners = this.partners.filter(p => p.partner_id !== id)
      } catch (err: any) {
        throw err
      }
    }
  }
})
